File: flowcatalyst-typescript/packages/domain-core/src/domain-event.ts
/**
 * Domain Event Interface
 *
 * Base interface for all domain events. Domain events represent facts about
 * what happened in the domain (past tense). Each event has its own schema
 * and is stored in the event store.
 *
 * Events follow the CloudEvents specification structure with additional
 * fields for tracing and ordering.
 *
 * Naming convention: Events are named in past tense describing what happened.
 * - `EventTypeCreated` (not CreateEventType)
 * - `SchemaFinalised` (not FinaliseSchema)
 * - `ApplicationActivated` (not ActivateApplication)
 *
 * Implementation note: Events should be implemented as readonly objects
 * for immutability.
 */

import { generate, type EntityTypeKey } from '@flowcatalyst/tsid';
import type { ExecutionContext } from './execution-context.js';

/**
 * Base interface for all domain events.
 */
export interface DomainEvent {
	/** Unique identifier for this event (TSID Crockford Base32 string) */
	readonly eventId: string;

	/**
	 * Event type code following the format: {app}:{domain}:{aggregate}:{action}
	 * Example: "platform:control-plane:eventtype:created"
	 */
	readonly eventType: string;

	/**
	 * Schema version of this event type (e.g., "1.0").
	 * Used for event versioning and schema evolution.
	 */
	readonly specVersion: string;

	/**
	 * Source system that generated this event.
	 * Example: "platform:control-plane"
	 */
	readonly source: string;

	/**
	 * Qualified aggregate identifier.
	 * Format: {domain}.{aggregate}.{id}
	 * Example: "platform.eventtype.0HZXEQ5Y8JY5Z"
	 */
	readonly subject: string;

	/** When the event occurred */
	readonly time: Date;

	/**
	 * Execution ID for tracking a single use case execution.
	 * All events from the same use case execution share this ID.
	 */
	readonly executionId: string;

	/**
	 * Correlation ID for distributed tracing.
	 * Typically the original request ID that started the chain.
	 */
	readonly correlationId: string;

	/**
	 * ID of the event that caused this event (if any).
	 * Used to build causal chains between events.
	 */
	readonly causationId: string | null;

	/** Principal who initiated the action that produced this event */
	readonly principalId: string;

	/**
	 * Message group for ordering guarantees.
	 * Events in the same message group are processed in order.
	 * Example: "platform:eventtype:0HZXEQ5Y8JY5Z"
	 */
	readonly messageGroup: string;

	/**
	 * Serialize the event-specific data payload to JSON.
	 * This contains the domain-specific fields of the event.
	 */
	toDataJson(): string;
}

/**
 * Base fields required to construct a domain event.
 */
export interface DomainEventBase {
	eventType: string;
	specVersion: string;
	source: string;
	subject: string;
	messageGroup: string;
}

/**
 * Helper to create domain event metadata from execution context.
 */
export interface DomainEventMetadata {
	eventId: string;
	executionId: string;
	correlationId: string;
	causationId: string | null;
	principalId: string;
	time: Date;
}

/**
 * DomainEvent factory helpers.
 */
export const DomainEvent = {
	/**
	 * Generate a new event ID.
	 */
	generateId(): string {
		return generate('EVENT');
	},

	/**
	 * Create event metadata from an execution context.
	 *
	 * @param ctx - The execution context
	 * @returns Event metadata
	 */
	metadataFrom(ctx: ExecutionContext): DomainEventMetadata {
		return {
			eventId: generate('EVENT'),
			executionId: ctx.executionId,
			correlationId: ctx.correlationId,
			causationId: ctx.causationId,
			principalId: ctx.principalId,
			time: new Date(),
		};
	},

	/**
	 * Create a subject string for an aggregate.
	 *
	 * @param domain - The domain name (e.g., "platform")
	 * @param aggregate - The aggregate type (e.g., "eventtype")
	 * @param id - The aggregate ID
	 * @returns The subject string (e.g., "platform.eventtype.0HZXEQ5Y8JY5Z")
	 */
	subject(domain: string, aggregate: string, id: string): string {
		return `${domain}.${aggregate}.${id}`;
	},

	/**
	 * Create a message group string for an aggregate.
	 *
	 * @param domain - The domain name (e.g., "platform")
	 * @param aggregate - The aggregate type (e.g., "eventtype")
	 * @param id - The aggregate ID
	 * @returns The message group string (e.g., "platform:eventtype:0HZXEQ5Y8JY5Z")
	 */
	messageGroup(domain: string, aggregate: string, id: string): string {
		return `${domain}:${aggregate}:${id}`;
	},

	/**
	 * Create an event type code.
	 *
	 * @param app - The application name (e.g., "platform")
	 * @param domain - The domain name (e.g., "control-plane")
	 * @param aggregate - The aggregate type (e.g., "eventtype")
	 * @param action - The action in past tense (e.g., "created")
	 * @returns The event type code (e.g., "platform:control-plane:eventtype:created")
	 */
	eventType(app: string, domain: string, aggregate: string, action: string): string {
		return `${app}:${domain}:${aggregate}:${action}`;
	},

	/**
	 * Extract the aggregate type from a subject string.
	 *
	 * Subject format: "platform.eventtype.123456789"
	 * Returns: "Eventtype" (capitalized)
	 */
	extractAggregateType(subject: string): string {
		if (!subject) {
			return 'Unknown';
		}
		const parts = subject.split('.');
		if (parts.length >= 2) {
			const aggregateType = parts[1]!;
			return aggregateType.charAt(0).toUpperCase() + aggregateType.slice(1).replace(/-/g, '');
		}
		return 'Unknown';
	},

	/**
	 * Extract the entity ID from a subject string.
	 *
	 * Subject format: "platform.eventtype.123456789"
	 * Returns: "123456789"
	 */
	extractEntityId(subject: string): string | null {
		if (!subject) {
			return null;
		}
		const parts = subject.split('.');
		if (parts.length >= 3) {
			return parts[2]!;
		}
		return null;
	},
};

/**
 * Abstract base class for domain events that provides common functionality.
 *
 * Extend this class to create concrete domain events:
 *
 * @example
 * ```typescript
 * interface EventTypeCreatedData {
 *     eventTypeId: string;
 *     code: string;
 *     name: string;
 * }
 *
 * class EventTypeCreated extends BaseDomainEvent<EventTypeCreatedData> {
 *     constructor(ctx: ExecutionContext, data: EventTypeCreatedData) {
 *         super({
 *             eventType: 'platform:control-plane:eventtype:created',
 *             specVersion: '1.0',
 *             source: 'platform:control-plane',
 *             subject: DomainEvent.subject('platform', 'eventtype', data.eventTypeId),
 *             messageGroup: DomainEvent.messageGroup('platform', 'eventtype', data.eventTypeId),
 *         }, ctx, data);
 *     }
 * }
 * ```
 */
export abstract class BaseDomainEvent<TData extends Record<string, unknown>> implements DomainEvent {
	readonly eventId: string;
	readonly eventType: string;
	readonly specVersion: string;
	readonly source: string;
	readonly subject: string;
	readonly time: Date;
	readonly executionId: string;
	readonly correlationId: string;
	readonly causationId: string | null;
	readonly principalId: string;
	readonly messageGroup: string;

	protected readonly data: TData;

	constructor(base: DomainEventBase, ctx: ExecutionContext, data: TData) {
		const metadata = DomainEvent.metadataFrom(ctx);

		this.eventId = metadata.eventId;
		this.eventType = base.eventType;
		this.specVersion = base.specVersion;
		this.source = base.source;
		this.subject = base.subject;
		this.time = metadata.time;
		this.executionId = metadata.executionId;
		this.correlationId = metadata.correlationId;
		this.causationId = metadata.causationId;
		this.principalId = metadata.principalId;
		this.messageGroup = base.messageGroup;
		this.data = data;
	}

	toDataJson(): string {
		return JSON.stringify(this.data);
	}

	/**
	 * Get the event-specific data payload.
	 */
	getData(): TData {
		return this.data;
	}
}

File: flowcatalyst-go/internal/platform/common/domain_event.go
package common

import (
	"encoding/json"
	"time"

	"go.flowcatalyst.tech/internal/common/tsid"
)

// DomainEvent is the interface that all domain events must implement.
// It follows the CloudEvents specification for interoperability.
//
// Domain events represent something that happened in the domain.
// They are immutable facts that are persisted for event sourcing
// and audit trail purposes.
type DomainEvent interface {
	// EventID returns the unique identifier for this event.
	// Format: TSID Crockford Base32 string.
	EventID() string

	// EventType returns the type code for this event.
	// Format: {app}:{domain}:{aggregate}:{action}
	// Example: "platform:control-plane:eventtype:created"
	EventType() string

	// SpecVersion returns the schema version of this event type.
	// Example: "1.0"
	SpecVersion() string

	// Source returns the system that generated this event.
	// Example: "platform:control-plane"
	Source() string

	// Subject returns the qualified aggregate identifier.
	// Format: {domain}.{aggregate}.{id}
	// Example: "platform.eventtype.0HZXEQ5Y8JY5Z"
	Subject() string

	// Time returns when the event occurred.
	Time() time.Time

	// CorrelationID returns the distributed tracing identifier.
	CorrelationID() string

	// CausationID returns the ID of the event that caused this event.
	// May be empty if this is a root event.
	CausationID() string

	// ExecutionID returns the unique ID for this use case execution.
	// Different from EventID - ExecutionID groups all events from one execution.
	ExecutionID() string

	// PrincipalID returns the ID of who initiated the action.
	PrincipalID() string

	// MessageGroup returns the group key for FIFO ordering.
	// Messages with the same group are processed in order.
	// Example: "platform:eventtype:0HZXEQ5Y8JY5Z"
	MessageGroup() string

	// ToDataJSON serializes the event-specific payload to JSON.
	ToDataJSON() string
}

// BaseDomainEvent provides a base implementation of DomainEvent
// that can be embedded in concrete event types.
//
// Example usage:
//
//	type EventTypeCreated struct {
//	    common.BaseDomainEvent
//	    EventTypeID string `json:"eventTypeId"`
//	    Code        string `json:"code"`
//	    Name        string `json:"name"`
//	}
type BaseDomainEvent struct {
	ID          string    `json:"eventId" bson:"_id"`
	Type        string    `json:"eventType" bson:"type"`
	Version     string    `json:"specVersion" bson:"specVersion"`
	Src         string    `json:"source" bson:"source"`
	Subj        string    `json:"subject" bson:"subject"`
	Timestamp   time.Time `json:"time" bson:"time"`
	Correlation string    `json:"correlationId" bson:"correlationId"`
	Causation   string    `json:"causationId,omitempty" bson:"causationId,omitempty"`
	Execution   string    `json:"executionId" bson:"executionId"`
	Principal   string    `json:"principalId" bson:"principalId"`
	MsgGroup    string    `json:"messageGroup" bson:"messageGroup"`
}

// NewBaseDomainEvent creates a new BaseDomainEvent with fields populated
// from the execution context.
func NewBaseDomainEvent(
	ctx *ExecutionContext,
	eventType string,
	subject string,
	messageGroup string,
) BaseDomainEvent {
	return BaseDomainEvent{
		ID:          tsid.Generate(),
		Type:        eventType,
		Version:     "1.0",
		Src:         "platform:control-plane",
		Subj:        subject,
		Timestamp:   time.Now(),
		Correlation: ctx.CorrelationID,
		Causation:   ctx.CausationID,
		Execution:   ctx.ExecutionID,
		Principal:   ctx.PrincipalID,
		MsgGroup:    messageGroup,
	}
}

// NewBaseDomainEventWithVersion creates a BaseDomainEvent with a specific spec version.
func NewBaseDomainEventWithVersion(
	ctx *ExecutionContext,
	eventType string,
	specVersion string,
	subject string,
	messageGroup string,
) BaseDomainEvent {
	base := NewBaseDomainEvent(ctx, eventType, subject, messageGroup)
	base.Version = specVersion
	return base
}

// NewBaseDomainEventWithSource creates a BaseDomainEvent with a custom source.
func NewBaseDomainEventWithSource(
	ctx *ExecutionContext,
	eventType string,
	source string,
	subject string,
	messageGroup string,
) BaseDomainEvent {
	base := NewBaseDomainEvent(ctx, eventType, subject, messageGroup)
	base.Src = source
	return base
}

// DomainEvent interface implementation for BaseDomainEvent

func (e BaseDomainEvent) EventID() string       { return e.ID }
func (e BaseDomainEvent) EventType() string     { return e.Type }
func (e BaseDomainEvent) SpecVersion() string   { return e.Version }
func (e BaseDomainEvent) Source() string        { return e.Src }
func (e BaseDomainEvent) Subject() string       { return e.Subj }
func (e BaseDomainEvent) Time() time.Time       { return e.Timestamp }
func (e BaseDomainEvent) CorrelationID() string { return e.Correlation }
func (e BaseDomainEvent) CausationID() string   { return e.Causation }
func (e BaseDomainEvent) ExecutionID() string   { return e.Execution }
func (e BaseDomainEvent) PrincipalID() string   { return e.Principal }
func (e BaseDomainEvent) MessageGroup() string  { return e.MsgGroup }

// ToDataJSON returns an empty object for the base event.
// Concrete event types should override this to include their payload.
func (e BaseDomainEvent) ToDataJSON() string {
	return "{}"
}

// PersistedEvent represents a domain event as stored in MongoDB.
// It includes the full event structure plus searchable context data.
type PersistedEvent struct {
	ID              string        `bson:"_id" json:"id"`
	SpecVersion     string        `bson:"specVersion" json:"specVersion"`
	Type            string        `bson:"type" json:"type"`
	Source          string        `bson:"source" json:"source"`
	Subject         string        `bson:"subject" json:"subject"`
	Time            time.Time     `bson:"time" json:"time"`
	Data            string        `bson:"data" json:"data"`
	CorrelationID   string        `bson:"correlationId" json:"correlationId"`
	CausationID     string        `bson:"causationId,omitempty" json:"causationId,omitempty"`
	ExecutionID     string        `bson:"executionId" json:"executionId"`
	DeduplicationID string        `bson:"deduplicationId" json:"deduplicationId"`
	MessageGroup    string        `bson:"messageGroup" json:"messageGroup"`
	ContextData     []ContextData `bson:"contextData" json:"contextData"`
	ClientID        string        `bson:"clientId,omitempty" json:"clientId,omitempty"`
}

// ContextData represents searchable key-value metadata on an event.
type ContextData struct {
	Key   string `bson:"key" json:"key"`
	Value string `bson:"value" json:"value"`
}

// ToPersistedEvent converts a DomainEvent to a PersistedEvent for storage.
func ToPersistedEvent(event DomainEvent, clientID string) *PersistedEvent {
	contextData := []ContextData{
		{Key: "principalId", Value: event.PrincipalID()},
		{Key: "aggregateType", Value: extractAggregateType(event.Subject())},
	}

	return &PersistedEvent{
		ID:              event.EventID(),
		SpecVersion:     event.SpecVersion(),
		Type:            event.EventType(),
		Source:          event.Source(),
		Subject:         event.Subject(),
		Time:            event.Time(),
		Data:            event.ToDataJSON(),
		CorrelationID:   event.CorrelationID(),
		CausationID:     event.CausationID(),
		ExecutionID:     event.ExecutionID(),
		DeduplicationID: event.EventType() + "-" + event.EventID(),
		MessageGroup:    event.MessageGroup(),
		ContextData:     contextData,
		ClientID:        clientID,
	}
}

// extractAggregateType extracts the aggregate type from a subject string.
// Subject format: {domain}.{aggregate}.{id}
// Example: "platform.eventtype.0HZXEQ5Y8JY5Z" -> "eventtype"
func extractAggregateType(subject string) string {
	// Simple extraction - find the second segment
	start := 0
	dotCount := 0
	for i, c := range subject {
		if c == '.' {
			dotCount++
			if dotCount == 1 {
				start = i + 1
			} else if dotCount == 2 {
				return subject[start:i]
			}
		}
	}
	if dotCount == 1 {
		return subject[start:]
	}
	return subject
}

// MarshalDataJSON is a helper to serialize event payload to JSON.
func MarshalDataJSON(data any) string {
	bytes, err := json.Marshal(data)
	if err != nil {
		return "{}"
	}
	return string(bytes)
}
